mne/epochs.py

```diff
@@ -39,39 +37,87 @@
 
         self.verbose = verbose
         self.name = name
 
-        if isinstance(event_id, dict):
+        if on_missing not in ['error', 'warning', 'ignore']:
+            raise ValueError('on_missing must be one of: error, '
+                             'warning, ignore. Got: %s' % on_missing)
+
+        # check out event_id dict
+        if event_id is None:  # convert to int to make typing-checks happy
+            event_id = dict((str(e), int(e)) for e in np.unique(events[:, 2]))
+        elif isinstance(event_id, dict):
             if not all(isinstance(v, int) for v in event_id.values()):
                 raise ValueError('Event IDs must be of type integer')
             if not all(isinstance(k, string_types) for k in event_id):
                 raise ValueError('Event names must be of type str')
-            self.event_id = event_id
         elif isinstance(event_id, list):
             if not all(isinstance(v, int) for v in event_id):
                 raise ValueError('Event IDs must be of type integer')
-            self.event_id = dict(zip((str(i) for i in event_id), event_id))
+            event_id = dict(zip((str(i) for i in event_id), event_id))
         elif isinstance(event_id, int):
-            self.event_id = {str(event_id): event_id}
+            event_id = {str(event_id): event_id}
         else:
             raise ValueError('event_id must be dict or int.')
+        self.event_id = event_id
+        del event_id
+
+        if events is not None:  # RtEpochs can have events=None
+            for key, val in self.event_id.items():
+                if val not in events[:, 2]:
+                    msg = ('No matching events found for %s '
+                           '(event id %i)' % (key, val))
+                    if on_missing == 'error':
+                        raise ValueError(msg)
+                    elif on_missing == 'warning':
+                        logger.warning(msg)
+                        warnings.warn(msg)
+                    else:  # on_missing == 'ignore':
+                        pass
+
+            values = list(self.event_id.values())
+            selected = in1d(events[:, 2], values)
+            self.selection = np.where(selected)[0]
+            self.drop_log = [list() if k in self.selection else ['IGNORED']
+                             for k in range(len(events))]
+            events = events[selected]
+            n_events = len(events)
+            if n_events > 1:
+                if np.diff(events.astype(np.int64)[:, 0]).min() <= 0:
+                    warnings.warn('The events passed to the Epochs '
+                                  'constructor are not chronologically '
+                                  'ordered.', RuntimeWarning)
+
+            if n_events > 0:
+                logger.info('%d matching events found' % n_events)
+            else:
+                raise ValueError('No desired events found.')
+            self.events = events
+            del events
```

Comment:
It's a safety thing -- by removing it from the namespace we ensure that
from that point forward we only make use of self.events